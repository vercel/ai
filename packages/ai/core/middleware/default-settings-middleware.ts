import {
  LanguageModelV1CallOptions,
  LanguageModelV1ProviderMetadata,
} from '@ai-sdk/provider';
import type { LanguageModelV1Middleware } from './language-model-v1-middleware';
import { mergeObjects } from '../util/merge-objects';

/**
 * Applies default settings for a language model.
 */
export function defaultSettingsMiddleware({
  settings,
}: {
  settings: Partial<
    LanguageModelV1CallOptions & {
      providerMetadata?: LanguageModelV1ProviderMetadata;
    }
  >;
}): LanguageModelV1Middleware {
  return {
    middlewareVersion: 'v1',
    transformParams: async ({ params }) => {
      return {
        ...settings,
        ...params,
        providerMetadata: mergeObjects(
          settings.providerMetadata,
          params.providerMetadata,
        ),

        // special case for temperature 0
        // TODO remove when temperature defaults to undefined
        temperature:
          params.temperature === 0 || params.temperature == null
            ? (settings.temperature ?? 0)
            : params.temperature,
      };
    },
  };
}
