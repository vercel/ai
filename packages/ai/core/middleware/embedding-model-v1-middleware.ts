import {
  EmbeddingModelV1,
  EmbeddingModelV1CallOptions,
} from '@ai-sdk/provider';

/**
 * Middleware for EmbeddingModelV1.
 * This type defines the structure for middleware that can be used to modify
 * the behavior of EmbeddingModelV1 operations.
 */
export type EmbeddingModelV1Middleware<VALUE> = {
  /**
   * Middleware specification version. Use `v1` for the current version.
   */
  middlewareVersion?: 'v1' | undefined; // backwards compatibility

  /**
   * Transforms the parameters before they are passed to the embedding model.
   * @param options - Object containing the parameters.
   * @param options.params - The original parameters for the embedding model call.
   * @returns A promise that resolves to the transformed parameters.
   */
  transformParams?: (options: {
    params: EmbeddingModelV1CallOptions<VALUE>;
  }) => PromiseLike<EmbeddingModelV1CallOptions<VALUE>>;

  /**
   * Wraps the embed operation of the embedding model.
   * @param options - Object containing the embed function, parameters, and model.
   * @param options.doEmbed - The original embed function.
   * @param options.params - The parameters for the embed call. If the
   * `transformParams` middleware is used, this will be the transformed parameters.
   * @param options.model - The embedding model instance.
   * @returns A promise that resolves to the result of the embed operation.
   */
  wrapEmbed?: (options: {
    doEmbed: () => ReturnType<EmbeddingModelV1<VALUE>['doEmbed']>;
    params: EmbeddingModelV1CallOptions<VALUE>;
    model: EmbeddingModelV1<VALUE>;
  }) => Promise<Awaited<ReturnType<EmbeddingModelV1<VALUE>['doEmbed']>>>;
};
