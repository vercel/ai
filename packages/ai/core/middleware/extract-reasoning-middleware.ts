import type {
  LanguageModelV2Content,
  LanguageModelV2Middleware,
  LanguageModelV2StreamPart,
} from '@ai-sdk/provider';
import { getPotentialStartIndex } from '../util/get-potential-start-index';

/**
 * Extract an XML-tagged reasoning section from the generated text and exposes it
 * as a `reasoning` property on the result.
 *
 * @param tagName - The name of the XML tag to extract reasoning from.
 * @param separator - The separator to use between reasoning and text sections.
 * @param startWithReasoning - Whether to start with reasoning tokens.
 */
export function extractReasoningMiddleware({
  tagName,
  separator = '\n',
  startWithReasoning = false,
}: {
  tagName: string;
  separator?: string;
  startWithReasoning?: boolean;
}): LanguageModelV2Middleware {
  const openingTag = `<${tagName}>`;
  const closingTag = `<\/${tagName}>`;

  return {
    middlewareVersion: 'v2',
    wrapGenerate: async ({ doGenerate }) => {
      const { content, ...rest } = await doGenerate();

      const transformedContent: LanguageModelV2Content[] = [];
      for (const part of content) {
        if (part.type !== 'text') {
          transformedContent.push(part);
          continue;
        }

        const text = startWithReasoning ? openingTag + part.text : part.text;

        const regexp = new RegExp(`${openingTag}(.*?)${closingTag}`, 'gs');
        const matches = Array.from(text.matchAll(regexp));

        if (!matches.length) {
          transformedContent.push(part);
          continue;
        }

        const reasoningText = matches.map(match => match[1]).join(separator);

        let textWithoutReasoning = text;
        for (let i = matches.length - 1; i >= 0; i--) {
          const match = matches[i];

          const beforeMatch = textWithoutReasoning.slice(0, match.index);
          const afterMatch = textWithoutReasoning.slice(
            match.index! + match[0].length,
          );

          textWithoutReasoning =
            beforeMatch +
            (beforeMatch.length > 0 && afterMatch.length > 0 ? separator : '') +
            afterMatch;
        }

        transformedContent.push({
          type: 'reasoning',
          reasoningType: 'text',
          text: reasoningText,
        });

        transformedContent.push({
          type: 'text',
          text: textWithoutReasoning,
        });
      }

      return { content: transformedContent, ...rest };
    },

    wrapStream: async ({ doStream }) => {
      const { stream, ...rest } = await doStream();

      let isFirstReasoning = true;
      let isFirstText = true;
      let afterSwitch = false;
      let isReasoning = startWithReasoning;
      let buffer = '';

      return {
        stream: stream.pipeThrough(
          new TransformStream<
            LanguageModelV2StreamPart,
            LanguageModelV2StreamPart
          >({
            transform: (chunk, controller) => {
              if (chunk.type !== 'text') {
                controller.enqueue(chunk);
                return;
              }

              buffer += chunk.text;

              function publish(text: string) {
                if (text.length > 0) {
                  const prefix =
                    afterSwitch &&
                    (isReasoning ? !isFirstReasoning : !isFirstText)
                      ? separator
                      : '';

                  controller.enqueue(
                    isReasoning
                      ? {
                          type: 'reasoning',
                          reasoningType: 'text',
                          text: prefix + text,
                        }
                      : {
                          type: 'text',
                          text: prefix + text,
                        },
                  );
                  afterSwitch = false;

                  if (isReasoning) {
                    isFirstReasoning = false;
                  } else {
                    isFirstText = false;
                  }
                }
              }

              do {
                const nextTag = isReasoning ? closingTag : openingTag;
                const startIndex = getPotentialStartIndex(buffer, nextTag);

                // no opening or closing tag found, publish the buffer
                if (startIndex == null) {
                  publish(buffer);
                  buffer = '';
                  break;
                }

                // publish text before the tag
                publish(buffer.slice(0, startIndex));

                const foundFullMatch =
                  startIndex + nextTag.length <= buffer.length;

                if (foundFullMatch) {
                  buffer = buffer.slice(startIndex + nextTag.length);
                  isReasoning = !isReasoning;
                  afterSwitch = true;
                } else {
                  buffer = buffer.slice(startIndex);
                  break;
                }
              } while (true);
            },
          }),
        ),
        ...rest,
      };
    },
  };
}
