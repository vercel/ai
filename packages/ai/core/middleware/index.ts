export { defaultSettingsMiddleware } from './default-settings-middleware';
export { extractReasoningMiddleware } from './extract-reasoning-middleware';
export type {
  Experimental_LanguageModelV1Middleware,
  LanguageModelV1Middleware,
} from './language-model-v1-middleware';
export { simulateStreamingMiddleware } from './simulate-streaming-middleware';
export {
  experimental_wrapLanguageModel,
  wrapLanguageModel,
} from './wrap-language-model';
