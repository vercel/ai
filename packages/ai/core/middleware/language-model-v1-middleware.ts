import { LanguageModelV1, LanguageModelV1CallOptions } from '@ai-sdk/provider';

/**
 * Experimental middleware for LanguageModelV1.
 * This type defines the structure for middleware that can be used to modify
 * the behavior of LanguageModelV1 operations.
 */
export type LanguageModelV1Middleware = {
  /**
   * Middleware specification version. Use `v1` for the current version.
   */
  middlewareVersion?: 'v1' | undefined; // backwards compatibility

  /**
   * Transforms the parameters before they are passed to the language model.
   * @param options - Object containing the type of operation and the parameters.
   * @param options.type - The type of operation ('generate' or 'stream').
   * @param options.params - The original parameters for the language model call.
   * @returns A promise that resolves to the transformed parameters.
   */
  transformParams?: (options: {
    type: 'generate' | 'stream';
    params: LanguageModelV1CallOptions;
  }) => PromiseLike<LanguageModelV1CallOptions>;

  /**
   * Wraps the generate operation of the language model.
   * @param options - Object containing the generate function, parameters, and model.
   * @param options.doGenerate - The original generate function.
   * @param options.doStream - The original stream function.
   * @param options.params - The parameters for the generate call. If the
   * `transformParams` middleware is used, this will be the transformed parameters.
   * @param options.model - The language model instance.
   * @returns A promise that resolves to the result of the generate operation.
   */
  wrapGenerate?: (options: {
    doGenerate: () => ReturnType<LanguageModelV1['doGenerate']>;
    doStream: () => ReturnType<LanguageModelV1['doStream']>;
    params: LanguageModelV1CallOptions;
    model: LanguageModelV1;
  }) => Promise<Awaited<ReturnType<LanguageModelV1['doGenerate']>>>;

  /**
   * Wraps the stream operation of the language model.
   *
   * @param options - Object containing the stream function, parameters, and model.
   * @param options.doGenerate - The original generate function.
   * @param options.doStream - The original stream function.
   * @param options.params - The parameters for the stream call. If the
   * `transformParams` middleware is used, this will be the transformed parameters.
   * @param options.model - The language model instance.
   * @returns A promise that resolves to the result of the stream operation.
   */
  wrapStream?: (options: {
    doGenerate: () => ReturnType<LanguageModelV1['doGenerate']>;
    doStream: () => ReturnType<LanguageModelV1['doStream']>;
    params: LanguageModelV1CallOptions;
    model: LanguageModelV1;
  }) => PromiseLike<Awaited<ReturnType<LanguageModelV1['doStream']>>>;
};

/**
 * @deprecated Use `LanguageModelV1Middleware` instead.
 */
// TODO remove in v5
export type Experimental_LanguageModelV1Middleware = LanguageModelV1Middleware;
