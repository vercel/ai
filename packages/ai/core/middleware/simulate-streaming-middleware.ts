import type { LanguageModelV1StreamPart } from '@ai-sdk/provider';
import type { LanguageModelV1Middleware } from './language-model-v1-middleware';

/**
 * Simulates streaming chunks with the response from a generate call.
 */
export function simulateStreamingMiddleware(): LanguageModelV1Middleware {
  return {
    middlewareVersion: 'v1',
    wrapStream: async ({ doGenerate }) => {
      const result = await doGenerate();

      const simulatedStream = new ReadableStream<LanguageModelV1StreamPart>({
        start(controller) {
          controller.enqueue({ type: 'response-metadata', ...result.response });

          if (result.reasoning) {
            if (typeof result.reasoning === 'string') {
              controller.enqueue({
                type: 'reasoning',
                textDelta: result.reasoning,
              });
            } else {
              for (const reasoning of result.reasoning) {
                switch (reasoning.type) {
                  case 'text': {
                    controller.enqueue({
                      type: 'reasoning',
                      textDelta: reasoning.text,
                    });
                    if (reasoning.signature != null) {
                      controller.enqueue({
                        type: 'reasoning-signature',
                        signature: reasoning.signature,
                      });
                    }
                    break;
                  }
                  case 'redacted': {
                    controller.enqueue({
                      type: 'redacted-reasoning',
                      data: reasoning.data,
                    });
                    break;
                  }
                }
              }
            }
          }

          if (result.text) {
            controller.enqueue({
              type: 'text-delta',
              textDelta: result.text,
            });
          }

          if (result.toolCalls) {
            for (const toolCall of result.toolCalls) {
              controller.enqueue({
                type: 'tool-call-delta',
                toolCallType: 'function',
                toolCallId: toolCall.toolCallId,
                toolName: toolCall.toolName,
                argsTextDelta: toolCall.args,
              });

              controller.enqueue({
                type: 'tool-call',
                ...toolCall,
              });
            }
          }

          controller.enqueue({
            type: 'finish',
            finishReason: result.finishReason,
            usage: result.usage,
            logprobs: result.logprobs,
            providerMetadata: result.providerMetadata,
          });

          controller.close();
        },
      });

      return {
        stream: simulatedStream,
        rawCall: result.rawCall,
        rawResponse: result.rawResponse,
        warnings: result.warnings,
      };
    },
  };
}
