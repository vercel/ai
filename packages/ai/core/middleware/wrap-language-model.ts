import { LanguageModelV1, LanguageModelV1CallOptions } from '@ai-sdk/provider';
import { LanguageModelV1Middleware } from './language-model-v1-middleware';
import { asArray } from '../../util/as-array';

/**
 * Wraps a LanguageModelV1 instance with middleware functionality.
 * This function allows you to apply middleware to transform parameters,
 * wrap generate operations, and wrap stream operations of a language model.
 *
 * @param options - Configuration options for wrapping the language model.
 * @param options.model - The original LanguageModelV1 instance to be wrapped.
 * @param options.middleware - The middleware to be applied to the language model. When multiple middlewares are provided, the first middleware will transform the input first, and the last middleware will be wrapped directly around the model.
 * @param options.modelId - Optional custom model ID to override the original model's ID.
 * @param options.providerId - Optional custom provider ID to override the original model's provider.
 * @returns A new LanguageModelV1 instance with middleware applied.
 */
export const wrapLanguageModel = ({
  model,
  middleware: middlewareArg,
  modelId,
  providerId,
}: {
  model: LanguageModelV1;
  middleware: LanguageModelV1Middleware | LanguageModelV1Middleware[];
  modelId?: string;
  providerId?: string;
}): LanguageModelV1 => {
  return asArray(middlewareArg)
    .reverse()
    .reduce((wrappedModel, middleware) => {
      return doWrap({ model: wrappedModel, middleware, modelId, providerId });
    }, model);
};

const doWrap = ({
  model,
  middleware: { transformParams, wrapGenerate, wrapStream },
  modelId,
  providerId,
}: {
  model: LanguageModelV1;
  middleware: LanguageModelV1Middleware;
  modelId?: string;
  providerId?: string;
}): LanguageModelV1 => {
  async function doTransform({
    params,
    type,
  }: {
    params: LanguageModelV1CallOptions;
    type: 'generate' | 'stream';
  }) {
    return transformParams ? await transformParams({ params, type }) : params;
  }

  return {
    specificationVersion: 'v1',

    provider: providerId ?? model.provider,
    modelId: modelId ?? model.modelId,

    defaultObjectGenerationMode: model.defaultObjectGenerationMode,
    supportsImageUrls: model.supportsImageUrls,
    supportsUrl: model.supportsUrl?.bind(model),
    supportsStructuredOutputs: model.supportsStructuredOutputs,

    async doGenerate(
      params: LanguageModelV1CallOptions,
    ): Promise<Awaited<ReturnType<LanguageModelV1['doGenerate']>>> {
      const transformedParams = await doTransform({ params, type: 'generate' });
      const doGenerate = async () => model.doGenerate(transformedParams);
      const doStream = async () => model.doStream(transformedParams);
      return wrapGenerate
        ? wrapGenerate({
            doGenerate,
            doStream,
            params: transformedParams,
            model,
          })
        : doGenerate();
    },

    async doStream(
      params: LanguageModelV1CallOptions,
    ): Promise<Awaited<ReturnType<LanguageModelV1['doStream']>>> {
      const transformedParams = await doTransform({ params, type: 'stream' });
      const doGenerate = async () => model.doGenerate(transformedParams);
      const doStream = async () => model.doStream(transformedParams);
      return wrapStream
        ? wrapStream({ doGenerate, doStream, params: transformedParams, model })
        : doStream();
    },
  };
};

/**
 * @deprecated Use `wrapLanguageModel` instead.
 */
// TODO remove in v5
export const experimental_wrapLanguageModel = wrapLanguageModel;
