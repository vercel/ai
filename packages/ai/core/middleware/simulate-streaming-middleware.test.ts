import {
  convertAsyncIterableToArray,
  mockId,
} from '@ai-sdk/provider-utils/test';
import { streamText } from '../generate-text';
import { wrapLanguageModel } from '../middleware/wrap-language-model';
import { MockLanguageModelV1 } from '../test/mock-language-model-v1';
import { simulateStreamingMiddleware } from './simulate-streaming-middleware';

const DEFAULT_SETTINGs = {
  prompt: 'Test prompt',
  experimental_generateMessageId: mockId({ prefix: 'msg' }),
  _internal: {
    generateId: mockId({ prefix: 'id' }),
    currentDate: () => new Date('2025-01-01'),
  },
};

describe('simulateStreamingMiddleware', () => {
  it('should simulate streaming with text response', async () => {
    const mockModel = new MockLanguageModelV1({
      async doGenerate() {
        return {
          text: 'This is a test response',
          finishReason: 'stop',
          usage: { promptTokens: 10, completionTokens: 10 },
          rawCall: { rawPrompt: '', rawSettings: {} },
        };
      },
    });

    const result = streamText({
      model: wrapLanguageModel({
        model: mockModel,
        middleware: simulateStreamingMiddleware(),
      }),
      ...DEFAULT_SETTINGs,
    });

    expect(
      await convertAsyncIterableToArray(result.fullStream),
    ).toMatchSnapshot();
  });

  it('should simulate streaming with reasoning as string', async () => {
    const mockModel = new MockLanguageModelV1({
      async doGenerate() {
        return {
          text: 'This is a test response',
          reasoning: 'This is the reasoning process',
          finishReason: 'stop',
          usage: { promptTokens: 10, completionTokens: 10 },
          rawCall: { rawPrompt: '', rawSettings: {} },
        };
      },
    });

    const result = streamText({
      model: wrapLanguageModel({
        model: mockModel,
        middleware: simulateStreamingMiddleware(),
      }),
      ...DEFAULT_SETTINGs,
    });

    expect(
      await convertAsyncIterableToArray(result.fullStream),
    ).toMatchSnapshot();
  });

  it('should simulate streaming with reasoning as array of text objects', async () => {
    const mockModel = new MockLanguageModelV1({
      async doGenerate() {
        return {
          text: 'This is a test response',
          reasoning: [
            { type: 'text', text: 'First reasoning step' },
            { type: 'text', text: 'Second reasoning step', signature: 'abc' },
          ],
          finishReason: 'stop',
          usage: { promptTokens: 10, completionTokens: 10 },
          rawCall: { rawPrompt: '', rawSettings: {} },
        };
      },
    });

    const result = streamText({
      model: wrapLanguageModel({
        model: mockModel,
        middleware: simulateStreamingMiddleware(),
      }),
      ...DEFAULT_SETTINGs,
    });

    expect(
      await convertAsyncIterableToArray(result.fullStream),
    ).toMatchSnapshot();
  });

  it('should simulate streaming with reasoning as array of mixed objects', async () => {
    const mockModel = new MockLanguageModelV1({
      async doGenerate() {
        return {
          text: 'This is a test response',
          reasoning: [
            { type: 'text', text: 'First reasoning step' },
            { type: 'redacted', data: 'data' },
          ],
          finishReason: 'stop',
          usage: { promptTokens: 10, completionTokens: 10 },
          rawCall: { rawPrompt: '', rawSettings: {} },
        };
      },
    });

    const result = streamText({
      model: wrapLanguageModel({
        model: mockModel,
        middleware: simulateStreamingMiddleware(),
      }),
      ...DEFAULT_SETTINGs,
    });

    expect(
      await convertAsyncIterableToArray(result.fullStream),
    ).toMatchSnapshot();
  });

  it('should simulate streaming with tool calls', async () => {
    const mockModel = new MockLanguageModelV1({
      async doGenerate() {
        return {
          text: 'This is a test response',
          toolCalls: [
            {
              toolCallId: 'tool-1',
              toolName: 'calculator',
              args: '{"expression": "2+2"}',
              toolCallType: 'function',
            },
            {
              toolCallId: 'tool-2',
              toolName: 'weather',
              args: '{"location": "New York"}',
              toolCallType: 'function',
            },
          ],
          finishReason: 'tool-calls',
          usage: { promptTokens: 10, completionTokens: 10 },
          rawCall: { rawPrompt: '', rawSettings: {} },
        };
      },
    });

    const result = streamText({
      model: wrapLanguageModel({
        model: mockModel,
        middleware: simulateStreamingMiddleware(),
      }),
      ...DEFAULT_SETTINGs,
    });

    expect(
      await convertAsyncIterableToArray(result.fullStream),
    ).toMatchSnapshot();
  });

  it('should preserve additional metadata in the response', async () => {
    const mockModel = new MockLanguageModelV1({
      async doGenerate() {
        return {
          text: 'This is a test response',
          finishReason: 'stop',
          usage: { promptTokens: 10, completionTokens: 10 },
          rawCall: { rawPrompt: '', rawSettings: {} },
          providerMetadata: { custom: { key: 'value' } },
        };
      },
    });

    const result = streamText({
      model: wrapLanguageModel({
        model: mockModel,
        middleware: simulateStreamingMiddleware(),
      }),
      ...DEFAULT_SETTINGs,
    });

    expect(
      await convertAsyncIterableToArray(result.fullStream),
    ).toMatchSnapshot();
  });

  it('should handle empty text response', async () => {
    const mockModel = new MockLanguageModelV1({
      async doGenerate() {
        return {
          text: '',
          finishReason: 'stop',
          usage: { promptTokens: 10, completionTokens: 0 },
          rawCall: { rawPrompt: '', rawSettings: {} },
        };
      },
    });

    const result = streamText({
      model: wrapLanguageModel({
        model: mockModel,
        middleware: simulateStreamingMiddleware(),
      }),
      ...DEFAULT_SETTINGs,
    });

    expect(
      await convertAsyncIterableToArray(result.fullStream),
    ).toMatchSnapshot();
  });

  it('should pass through warnings from the model', async () => {
    const mockModel = new MockLanguageModelV1({
      async doGenerate() {
        return {
          text: 'This is a test response',
          finishReason: 'stop',
          usage: { promptTokens: 10, completionTokens: 10 },
          rawCall: { rawPrompt: '', rawSettings: {} },
          warnings: [
            { type: 'other', message: 'Test warning', code: 'test_warning' },
          ],
        };
      },
    });

    const result = streamText({
      model: wrapLanguageModel({
        model: mockModel,
        middleware: simulateStreamingMiddleware(),
      }),
      ...DEFAULT_SETTINGs,
    });

    result.consumeStream();

    expect(await result.warnings).toMatchSnapshot();
  });
});
