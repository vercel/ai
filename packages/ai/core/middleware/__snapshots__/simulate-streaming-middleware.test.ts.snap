// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`simulateStreamingMiddleware > should handle empty text response 1`] = `
[
  {
    "messageId": "msg-6",
    "request": {},
    "type": "step-start",
    "warnings": [],
  },
  {
    "experimental_providerMetadata": undefined,
    "finishReason": "stop",
    "isContinued": false,
    "logprobs": undefined,
    "messageId": "msg-6",
    "providerMetadata": undefined,
    "request": {},
    "response": {
      "headers": undefined,
      "id": "id-13",
      "modelId": "mock-model-id",
      "timestamp": 2025-01-01T00:00:00.000Z,
    },
    "type": "step-finish",
    "usage": {
      "completionTokens": 0,
      "promptTokens": 10,
      "totalTokens": 10,
    },
    "warnings": undefined,
  },
  {
    "experimental_providerMetadata": undefined,
    "finishReason": "stop",
    "logprobs": undefined,
    "providerMetadata": undefined,
    "response": {
      "headers": undefined,
      "id": "id-13",
      "modelId": "mock-model-id",
      "timestamp": 2025-01-01T00:00:00.000Z,
    },
    "type": "finish",
    "usage": {
      "completionTokens": 0,
      "promptTokens": 10,
      "totalTokens": 10,
    },
  },
]
`;

exports[`simulateStreamingMiddleware > should pass through warnings from the model 1`] = `
[
  {
    "code": "test_warning",
    "message": "Test warning",
    "type": "other",
  },
]
`;

exports[`simulateStreamingMiddleware > should preserve additional metadata in the response 1`] = `
[
  {
    "messageId": "msg-5",
    "request": {},
    "type": "step-start",
    "warnings": [],
  },
  {
    "textDelta": "This is a test response",
    "type": "text-delta",
  },
  {
    "experimental_providerMetadata": {
      "custom": {
        "key": "value",
      },
    },
    "finishReason": "stop",
    "isContinued": false,
    "logprobs": undefined,
    "messageId": "msg-5",
    "providerMetadata": {
      "custom": {
        "key": "value",
      },
    },
    "request": {},
    "response": {
      "headers": undefined,
      "id": "id-11",
      "modelId": "mock-model-id",
      "timestamp": 2025-01-01T00:00:00.000Z,
    },
    "type": "step-finish",
    "usage": {
      "completionTokens": 10,
      "promptTokens": 10,
      "totalTokens": 20,
    },
    "warnings": undefined,
  },
  {
    "experimental_providerMetadata": {
      "custom": {
        "key": "value",
      },
    },
    "finishReason": "stop",
    "logprobs": undefined,
    "providerMetadata": {
      "custom": {
        "key": "value",
      },
    },
    "response": {
      "headers": undefined,
      "id": "id-11",
      "modelId": "mock-model-id",
      "timestamp": 2025-01-01T00:00:00.000Z,
    },
    "type": "finish",
    "usage": {
      "completionTokens": 10,
      "promptTokens": 10,
      "totalTokens": 20,
    },
  },
]
`;

exports[`simulateStreamingMiddleware > should simulate streaming with reasoning as array of mixed objects 1`] = `
[
  {
    "messageId": "msg-3",
    "request": {},
    "type": "step-start",
    "warnings": [],
  },
  {
    "textDelta": "First reasoning step",
    "type": "reasoning",
  },
  {
    "data": "data",
    "type": "redacted-reasoning",
  },
  {
    "textDelta": "This is a test response",
    "type": "text-delta",
  },
  {
    "experimental_providerMetadata": undefined,
    "finishReason": "stop",
    "isContinued": false,
    "logprobs": undefined,
    "messageId": "msg-3",
    "providerMetadata": undefined,
    "request": {},
    "response": {
      "headers": undefined,
      "id": "id-7",
      "modelId": "mock-model-id",
      "timestamp": 2025-01-01T00:00:00.000Z,
    },
    "type": "step-finish",
    "usage": {
      "completionTokens": 10,
      "promptTokens": 10,
      "totalTokens": 20,
    },
    "warnings": undefined,
  },
  {
    "experimental_providerMetadata": undefined,
    "finishReason": "stop",
    "logprobs": undefined,
    "providerMetadata": undefined,
    "response": {
      "headers": undefined,
      "id": "id-7",
      "modelId": "mock-model-id",
      "timestamp": 2025-01-01T00:00:00.000Z,
    },
    "type": "finish",
    "usage": {
      "completionTokens": 10,
      "promptTokens": 10,
      "totalTokens": 20,
    },
  },
]
`;

exports[`simulateStreamingMiddleware > should simulate streaming with reasoning as array of text objects 1`] = `
[
  {
    "messageId": "msg-2",
    "request": {},
    "type": "step-start",
    "warnings": [],
  },
  {
    "textDelta": "First reasoning step",
    "type": "reasoning",
  },
  {
    "textDelta": "Second reasoning step",
    "type": "reasoning",
  },
  {
    "signature": "abc",
    "type": "reasoning-signature",
  },
  {
    "textDelta": "This is a test response",
    "type": "text-delta",
  },
  {
    "experimental_providerMetadata": undefined,
    "finishReason": "stop",
    "isContinued": false,
    "logprobs": undefined,
    "messageId": "msg-2",
    "providerMetadata": undefined,
    "request": {},
    "response": {
      "headers": undefined,
      "id": "id-5",
      "modelId": "mock-model-id",
      "timestamp": 2025-01-01T00:00:00.000Z,
    },
    "type": "step-finish",
    "usage": {
      "completionTokens": 10,
      "promptTokens": 10,
      "totalTokens": 20,
    },
    "warnings": undefined,
  },
  {
    "experimental_providerMetadata": undefined,
    "finishReason": "stop",
    "logprobs": undefined,
    "providerMetadata": undefined,
    "response": {
      "headers": undefined,
      "id": "id-5",
      "modelId": "mock-model-id",
      "timestamp": 2025-01-01T00:00:00.000Z,
    },
    "type": "finish",
    "usage": {
      "completionTokens": 10,
      "promptTokens": 10,
      "totalTokens": 20,
    },
  },
]
`;

exports[`simulateStreamingMiddleware > should simulate streaming with reasoning as string 1`] = `
[
  {
    "messageId": "msg-1",
    "request": {},
    "type": "step-start",
    "warnings": [],
  },
  {
    "textDelta": "This is the reasoning process",
    "type": "reasoning",
  },
  {
    "textDelta": "This is a test response",
    "type": "text-delta",
  },
  {
    "experimental_providerMetadata": undefined,
    "finishReason": "stop",
    "isContinued": false,
    "logprobs": undefined,
    "messageId": "msg-1",
    "providerMetadata": undefined,
    "request": {},
    "response": {
      "headers": undefined,
      "id": "id-3",
      "modelId": "mock-model-id",
      "timestamp": 2025-01-01T00:00:00.000Z,
    },
    "type": "step-finish",
    "usage": {
      "completionTokens": 10,
      "promptTokens": 10,
      "totalTokens": 20,
    },
    "warnings": undefined,
  },
  {
    "experimental_providerMetadata": undefined,
    "finishReason": "stop",
    "logprobs": undefined,
    "providerMetadata": undefined,
    "response": {
      "headers": undefined,
      "id": "id-3",
      "modelId": "mock-model-id",
      "timestamp": 2025-01-01T00:00:00.000Z,
    },
    "type": "finish",
    "usage": {
      "completionTokens": 10,
      "promptTokens": 10,
      "totalTokens": 20,
    },
  },
]
`;

exports[`simulateStreamingMiddleware > should simulate streaming with text response 1`] = `
[
  {
    "messageId": "msg-0",
    "request": {},
    "type": "step-start",
    "warnings": [],
  },
  {
    "textDelta": "This is a test response",
    "type": "text-delta",
  },
  {
    "experimental_providerMetadata": undefined,
    "finishReason": "stop",
    "isContinued": false,
    "logprobs": undefined,
    "messageId": "msg-0",
    "providerMetadata": undefined,
    "request": {},
    "response": {
      "headers": undefined,
      "id": "id-1",
      "modelId": "mock-model-id",
      "timestamp": 2025-01-01T00:00:00.000Z,
    },
    "type": "step-finish",
    "usage": {
      "completionTokens": 10,
      "promptTokens": 10,
      "totalTokens": 20,
    },
    "warnings": undefined,
  },
  {
    "experimental_providerMetadata": undefined,
    "finishReason": "stop",
    "logprobs": undefined,
    "providerMetadata": undefined,
    "response": {
      "headers": undefined,
      "id": "id-1",
      "modelId": "mock-model-id",
      "timestamp": 2025-01-01T00:00:00.000Z,
    },
    "type": "finish",
    "usage": {
      "completionTokens": 10,
      "promptTokens": 10,
      "totalTokens": 20,
    },
  },
]
`;

exports[`simulateStreamingMiddleware > should simulate streaming with tool calls 1`] = `
[
  {
    "messageId": "msg-4",
    "request": {},
    "type": "step-start",
    "warnings": [],
  },
  {
    "textDelta": "This is a test response",
    "type": "text-delta",
  },
  {
    "error": [AI_NoSuchToolError: Model tried to call unavailable tool 'calculator'. No tools are available.],
    "type": "error",
  },
  {
    "error": [AI_NoSuchToolError: Model tried to call unavailable tool 'weather'. No tools are available.],
    "type": "error",
  },
  {
    "experimental_providerMetadata": undefined,
    "finishReason": "tool-calls",
    "isContinued": false,
    "logprobs": undefined,
    "messageId": "msg-4",
    "providerMetadata": undefined,
    "request": {},
    "response": {
      "headers": undefined,
      "id": "id-9",
      "modelId": "mock-model-id",
      "timestamp": 2025-01-01T00:00:00.000Z,
    },
    "type": "step-finish",
    "usage": {
      "completionTokens": 10,
      "promptTokens": 10,
      "totalTokens": 20,
    },
    "warnings": undefined,
  },
  {
    "experimental_providerMetadata": undefined,
    "finishReason": "tool-calls",
    "logprobs": undefined,
    "providerMetadata": undefined,
    "response": {
      "headers": undefined,
      "id": "id-9",
      "modelId": "mock-model-id",
      "timestamp": 2025-01-01T00:00:00.000Z,
    },
    "type": "finish",
    "usage": {
      "completionTokens": 10,
      "promptTokens": 10,
      "totalTokens": 20,
    },
  },
]
`;
